### PR TITLE
Starting bcf::IndexedReader

### PR DESCRIPTION
**Pull request by holtgrewe**
Partially adresses #68. However, only streaming over one file is currently implemented to give a `bcf::IndexedReader` interface.

Note that I somewhere made a mistake so the tests crash with an invalid memory access error.

I have to admit that I don't have an idea how to approach this with Rust tooling and would appreciate any help!

### CODE REVIEW

File: src/bcf/mod.rs [hunk 1]
```diff
@@ -92,10 +99,17 @@ impl Reader {
...
-    pub fn records(&mut self) -> Records {
+    fn records(&mut self) -> Records<Self> {
         Records { reader: self }
     }
+
+    fn header(&self) -> &HeaderView {
+        &self.header
+    }
+
+    fn empty_record(&self) -> Record {
```
Comment (johanneskoester) [inline src/bcf/mod.rs]: You should be able to move the implementation of empty_record into the Trait, by using the Read::header() method, right?

File: src/bcf/mod.rs [hunk 2]
```diff
@@ -108,6 +122,125 @@ impl Drop for Reader {
...
+    }
+}
+
+
+impl Read for IndexedReader {
+    fn read(&mut self, record: &mut record::Record) -> Result<(), ReadError> {
+        match unsafe { htslib::bcf_sr_next_line(self.inner) } {
+            0 => Err(ReadError::NoMoreRecord),
+            i => {
+                assert!(i > 0, "Must not be negative");
```
Comment (johanneskoester) [inline src/bcf/mod.rs]: Can't this be negative in case of an error when reading the file? I guess we would like to return a `ReadError` variant in such a case, rather than having a failing assertion.

File: src/bcf/mod.rs (hunk 1 above)
Comment (holtgrewe) [reply src/bcf/mod.rs]: OK

File: src/bcf/mod.rs (hunk 2 above)
Comment (holtgrewe) [reply src/bcf/mod.rs]: Actually, the code counts the number of readers for which a record is found... The API is a bit weird in this area.

**Comment by holtgrewe** [comment]
OK, pushed some fixes. I would propose to change the `bcf::Record` object as documented in the code at some later point when we have a more complete implementation of the synced reader and have a better feeling for that part of the API.

File: src/bcf/mod.rs (hunk 1 above)
Comment (holtgrewe) [reply src/bcf/mod.rs]: Arg, then I need to expose the `Rc<HeaderView>` member. Is this OK?

File: src/bcf/mod.rs (hunk 2 above)
Comment (johanneskoester) [reply src/bcf/mod.rs]: And how is an error indicated?

Comment (holtgrewe) [reply src/bcf/mod.rs]: I don't know. Here is the usage example from the docs:

```
    Example of usage:

        bcf_srs_t *sr = bcf_sr_init();
        bcf_sr_set_opt(sr, BCF_SR_PAIR_LOGIC, BCF_SR_PAIR_BOTH_REF);
        bcf_sr_set_opt(sr, BCF_SR_REQUIRE_IDX);
        for (i=0; i<nfiles; i++)
            bcf_sr_add_reader(sr,files[i]);
        while ( bcf_sr_next_line(sr) )
        {
            for (i=0; i<nfiles; i++)
            {
                bcf1_t *line = bcf_sr_get_line(sr,i);
                ...
            }
        }
        if ( sr->errnum ) error("Error: %s\n", bcf_sr_strerror(sr->errnum));
        bcf_sr_destroy(sr);
```

Comment (holtgrewe) [reply src/bcf/mod.rs]: gah, it's right there...

Comment (johanneskoester) [reply src/bcf/mod.rs]: Ah, looks like an error leads to termination of the loop (no more record) and then you can find an error code in `sr->errnum`? So, this means here that in the `0` case you would have to check `sr->errnunm` as well.

File: src/bcf/mod.rs [hunk 3]
```diff
@@ -22,10 +22,41 @@ pub use bcf::buffer::RecordBuffer;
...
+    ///
+    /// # Arguments
+    /// * record - an empty record, that can be created with `bcf::Reader::empty_record`.
+    fn read(&mut self, record: &mut record::Record) -> Result<(), ReadError>;
+
+    /// Return an iterator over all records of the VCF/BCF file.
+    fn records(&mut self) -> Records<Self>;
+
+    /// Return the header.
+    fn header(&self) -> Rc<HeaderView>;
```
Comment (johanneskoester) [inline src/bcf/mod.rs]: Why not returning `&HeaderView`?

Comment (johanneskoester) [reply src/bcf/mod.rs]: Ah, I think now I understand the problem with empty_record. In that case, I think it is better to keep the API clean, return &HeaderView, and leave the implementation of empty_record to the implementing structs.

Comment (holtgrewe) [reply src/bcf/mod.rs]: Because `bcf::record::Record::new` wants a `Rc<HeaderView>`...

File: src/bcf/mod.rs [hunk 4]
```diff
@@ -132,6 +151,149 @@ impl Drop for Reader {
...
+                    htslib::bcf_copy(
+                        record.inner,
+                        *(*(*self.inner).readers.offset(0)).buffer.offset(0));
+                }
+
+                record.set_header(self.header.clone());
+
+                match self.current_region {
+                    Some((rid, _start, end)) => {
+                        if record.rid().is_some() && rid == record.rid().unwrap() &&
```
Comment (johanneskoester) [inline src/bcf/mod.rs]: One last thing. Why is this necessary? I thought the fetch already fetches the correct region?

Comment (holtgrewe) [reply src/bcf/mod.rs]: It only jumps to the start of the region...

Comment (johanneskoester) [reply src/bcf/mod.rs]: ok!
